File: backend/src/modules/gateway/session-identity/request-utils.ts
import type { GatewaySessionIdentityRequest } from './types.js'

export function normalizedGatewaySessionRequestPath(request: GatewaySessionIdentityRequest): string {
  const endpoint = request.originalUrl || request.path || ''
  const path = endpoint.split('?', 1)[0]?.trim().toLowerCase() ?? ''
  if (!path) return '/'
  const normalized = path.replace(/\/+$/, '').replace(/^\/v1(?=\/|internal:|$)/, '')
  if (!normalized) return '/'
  return normalized.startsWith('/') ? normalized : `/${normalized}`
}

export function gatewaySessionHeaderValues(request: GatewaySessionIdentityRequest, name: string): string[] {
  const normalizedName = name.toLowerCase()
  const distinct = request.headersDistinct?.[normalizedName]
    ?? Object.entries(request.headersDistinct ?? {}).find(([key]) => key.toLowerCase() === normalizedName)?.[1]
  if (distinct?.length) {
    return distinct.map((value) => String(value))
  }
  const rawValues: string[] = []
  for (let index = 0; index + 1 < (request.rawHeaders?.length ?? 0); index += 2) {
    if (request.rawHeaders?.[index]?.toLowerCase() === normalizedName) {
      rawValues.push(String(request.rawHeaders[index + 1]))
    }
  }
  if (rawValues.length) return rawValues
  const direct = request.headers?.[normalizedName]
    ?? Object.entries(request.headers ?? {}).find(([key]) => key.toLowerCase() === normalizedName)?.[1]
  if (Array.isArray(direct)) {
    return direct.map((value) => String(value))
  }
  if (typeof direct === 'string') {
    return [direct]
  }
  const value = request.header?.(name)
  return typeof value === 'string' ? [value] : []
}

File: backend-go/projects/gateway/internal/gatewaysession/requestutils.go
package gatewaysession

import "strings"

// IdentityRequest mirrors the consumed surface of
// GatewaySessionIdentityRequest: the original URL / path plus multi-value
// header access (the headersDistinct projection).
type IdentityRequest interface {
	// OriginalURL mirrors request.originalUrl.
	OriginalURL() string
	// Path mirrors request.path.
	Path() string
	// HeaderValues mirrors request.headersDistinct[name]: every value sent
	// for the (case-insensitive) header name.
	HeaderValues(name string) []string
}

// NormalizedGatewaySessionRequestPath mirrors normalizedGatewaySessionRequestPath.
//
// The `/^\/v1(?=\/|internal:|$)/` lookahead is implemented manually (RE2 has
// no lookahead): the leading "/v1" is stripped only when followed by "/",
// "internal:" or the end of the string.
func NormalizedGatewaySessionRequestPath(request IdentityRequest) string {
	endpoint := ""
	if request != nil {
		endpoint = request.OriginalURL()
		if endpoint == "" {
			endpoint = request.Path()
		}
	}
	path := endpoint
	if idx := strings.IndexByte(endpoint, '?'); idx >= 0 {
		path = endpoint[:idx]
	}
	path = jsTrimString(strings.ToLower(path))
	if path == "" {
		return "/"
	}
	path = trimTrailingSlashes(path)
	path = stripV1Prefix(path)
	if path == "" {
		return "/"
	}
	if !strings.HasPrefix(path, "/") {
		return "/" + path
	}
	return path
}

func trimTrailingSlashes(path string) string {
	for strings.HasSuffix(path, "/") {
		path = path[:len(path)-1]
	}
	return path
}

func stripV1Prefix(path string) string {
	const prefix = "/v1"
	if !strings.HasPrefix(path, prefix) {
		return path
	}
	rest := path[len(prefix):]
	if rest == "" || strings.HasPrefix(rest, "/") || strings.HasPrefix(rest, "internal:") {
		return rest
	}
	return path
}

// GatewaySessionHeaderValues mirrors gatewaySessionHeaderValues: distinct
// header values first, then raw header pairs, then single-value headers.
// The IdentityRequest projection collapses the Node fallback chain into the
// headersDistinct accessor; resolvers only need the ordered value list.
func GatewaySessionHeaderValues(request IdentityRequest, name string) []string {
	if request == nil {
		return nil
	}
	values := request.HeaderValues(name)
	if len(values) == 0 {
		return nil
	}
	return values
}
